gateway-server/src/main/java/org/apache/knox/gateway/session/control/InMemoryConcurrentSessionVerifier.java

```diff
@@ -0,0 +1,169 @@
+  private int privilegedUserConcurrentSessionLimit;
+  private int nonPrivilegedUserConcurrentSessionLimit;
+  private Map<String, Set<SessionJWT>> concurrentSessionCounter;
+  private final Lock sessionCountModifyLock = new ReentrantLock();
+
+  @Override
+  public boolean verifySessionForUser(String username, JWT jwtToken) {
+    if (!privilegedUsers.contains(username) && !nonPrivilegedUsers.contains(username)) {
+      return true;
+    }
+
+    sessionCountModifyLock.lock();
+    try {
+      int validTokenNumber = countValidTokensForUser(username);
+      if (privilegedUserCheckLimitReached(username, validTokenNumber) || nonPrivilegedUserCheckLimitReached(username, validTokenNumber)) {
+        return false;
+      }
+      concurrentSessionCounter.putIfAbsent(username, new HashSet<>());
+      concurrentSessionCounter.compute(username, (key, sessionTokenSet) -> addTokenForUser(sessionTokenSet, jwtToken));
+    } finally {
+      sessionCountModifyLock.unlock();
+    }
+    return true;
+  }
+
+  private int countValidTokensForUser(String username) {
+    return (int) concurrentSessionCounter
+            .getOrDefault(username, Collections.emptySet())
+            .stream()
+            .filter(each -> !each.hasExpired())
+            .count();
+  }
+
+  private boolean privilegedUserCheckLimitReached(String username, int validTokenNumber) {
+    if (privilegedUserConcurrentSessionLimit < 0) {
+      return false;
+    }
+    return privilegedUsers.contains(username) && (validTokenNumber >= privilegedUserConcurrentSessionLimit);
+  }
+
+  private boolean nonPrivilegedUserCheckLimitReached(String username, int validTokenNumber) {
+    if (nonPrivilegedUserConcurrentSessionLimit < 0) {
+      return false;
+    }
+    return nonPrivilegedUsers.contains(username) && (validTokenNumber >= nonPrivilegedUserConcurrentSessionLimit);
+  }
+
+  @Override
+  public void sessionEndedForUser(String username, String token) {
+    if (StringUtils.isNotBlank(token)) {
+      sessionCountModifyLock.lock();
+      try {
+        concurrentSessionCounter.computeIfPresent(username, (key, sessionTokenSet) -> removeTokenFromUser(sessionTokenSet, token));
+      } finally {
+        sessionCountModifyLock.unlock();
+      }
+    }
+  }
+
+  private Set<SessionJWT> removeTokenFromUser(Set<SessionJWT> sessionTokenSet, String token) {
+    sessionTokenSet.removeIf(sessionToken -> sessionToken.getToken().equals(token));
+    if (sessionTokenSet.isEmpty()) {
+      return null;
+    }
+    return sessionTokenSet;
+  }
+
+  private Set<SessionJWT> addTokenForUser(Set<SessionJWT> sessionTokenSet, JWT jwtToken) {
+    sessionTokenSet.add(new SessionJWT(jwtToken));
+    return sessionTokenSet;
+  }
+
+  @Override
+  public void init(GatewayConfig config, Map<String, String> options) throws ServiceLifecycleException {
+    this.privilegedUsers = config.getPrivilegedUsers();
+    this.nonPrivilegedUsers = config.getNonPrivilegedUsers();
```

Comment:
Does this represent a configured set of non-privileged users for whom a concurrent session limit will be applied?
If so, why can't any user who is NOT configured as a member of the privileged users set be treated as a non-privileged user? Why do we need a separate configuration for these non-privileged users? What happens to those users who are in neither configured set? Do they get as many sessions as they want?

Reply:
Yes, there are two groups privileged and non-privileged, each has a limit. Users who are in neither of the groups can have unlimited sessions. @smolnar82 mentioned at the begging of the task that there might be overlapping between the groups so we need to have two groups configured, but if you suggest that configuring one group is enough, than there might have been some misunderstanding in this topic.

Reply:
https://issues.apache.org/jira/browse/KNOX-2789